### ISSUE TITLE
Incorrect constant folding behavior of floating-point operations

### ISSUE THREAD

**Issue by AaronBallman** [llvm/llvm-project#62479]
LLVM's constant folding appears to use the host system's math library to perform constant folding. e.g.,
https://github.com/llvm/llvm-project/blob/faa43a80f70c639f40021a175c60857bdcf7cc1f/llvm/lib/Analysis/ConstantFolding.cpp#L2407
(This is predominately a concern for cross-compilation scenarios where the host and target are different in interesting ways.)

What's more, it happily uses `double` to perform its operations, regardless of operand type. e.g.,
https://github.com/llvm/llvm-project/blob/faa43a80f70c639f40021a175c60857bdcf7cc1f/llvm/lib/Analysis/ConstantFolding.cpp#L1797
(This is predominately a concern about the use of `double` for all operations, regardless of source operand type, for native and cross compilation scenarios.)

This doesn't match conformance expectations for C. C2x 6.6p17 reads,
> The semantic rules for the evaluation of a constant expression are the same as for nonconstant expressions

while p5 reads,
> An expression that evaluates to a constant is required in several contexts. If a floating expression is evaluated in the translation environment, the arithmetic range and precision shall be at least as great as if the expression were being evaluated in the execution environment.

Use of `double` for performing the computation seems somewhat defensible per p5, but not for `long double` constant folding. It might be more defensible to use the host system's math library only when the host system floating-point model matches the target architecture's well enough to meet conformance requirements.

The requirements in C++ are a bit more relaxed than in C, but still recommend following the target architecture's semantics when folding at compile time: http://eel.is/c++draft/expr.const#14

**Comment by cor3ntin** [comment]
C++ had extensive discussions on this topic in the context of constexpr math functions.

I think the best way forward, and probably the only reasonable option as we most certainly don't want the front end to (try to) emulate all kind of uniquely quirky architecture is for the front end to have correctly rounded Floating point operations at compile time - when constant folding or constant evaluation.

I think we could achieve that by somehow factoring out the work done in llvm libc to implement all the correctly rounded math functions, and use that implementation in the front end.

https://libc.llvm.org/math/index.html#accuracy-of-higher-math-functions

This would, as a bonus make us consistent with what GCC is doing and it would make the result of constant evaluation correct, portable, and predictable.

It is possible that constexpr floating point functions will be considered for C++26 so there are more than one reason to consider that work.

**Comment by jcranmer-intel** [comment]
There's definitely a tension between "compile-time uses same function as runtime" and "compiler results are same as in cross-compiler;" it's impossible to choose both at the same time. Between these two options, I would prefer the latter one.

(It should also be noted that the existing code does not in fact even get us "compile-time uses the same function as runtime"--constant folding for `exp2` instead uses `pow(2.0, x)`, not to mention all the `float` implementations go through the `double` version instead.)

In the long run, I think the best option for us is to reuse correctly-rounded math functions from llvm-libc. I don't know how much work that would be, although it should also be noted that currently, only `float` versions are provided. I also think it's reasonable to exclude the non-IEEE types of `bfloat16`, `x86_fp80`, and `ppc_fp128` from any sort of constant folding (the latter two are just too cursed to be reasonable). Support for `fp128` is more reasonable than the other `long double` types, but excluding it on the basis of excluding all `long double` implementations is also reasonable to me, if perhaps less preferable.

**Comment by AaronBallman** [comment]
> I think last time I looked at this, I concluded we aren't actually breaking any conformance requirements; there isn't actually any requirement that calling a floating-point function with a given input consistently produces the same result. So for functions that aren't required to be correctly rounded, there are multiple possible results.

Can you cite something in the standard to back that assertion up? The two paragraphs I cited are pretty definitively stated, and C nails down the floating-point model more tightly than C++ does.

> > Use of double for performing the computation seems somewhat defensible per p5, but not for long double constant folding.
> 
> We don't fold "long double" arithmetic. There's a check that the type is "half", "float", or "double".

Ah, that's good at least!

**Comment by cor3ntin** [comment]
> > I think last time I looked at this, I concluded we aren't actually breaking any conformance requirements; there isn't actually any requirement that calling a floating-point function with a given input consistently produces the same result. So for functions that aren't required to be correctly rounded, there are multiple possible results.
> 
> Can you cite something in the standard to back that assertion up? The two paragraphs I cited are pretty definitively stated, and C nails down the floating-point model more tightly than C++ does.
> 

> arithmetic range and precision shall be at least as great as if the expression were being evaluated in the execution environment.

I don't think this is making a statement about value or correctness, is it? We just have to consider your same number of bits. At least that's how I read it.

**Comment by AaronBallman** [comment]
> > > I think last time I looked at this, I concluded we aren't actually breaking any conformance requirements; there isn't actually any requirement that calling a floating-point function with a given input consistently produces the same result. So for functions that aren't required to be correctly rounded, there are multiple possible results.
> > 
> > 
> > Can you cite something in the standard to back that assertion up? The two paragraphs I cited are pretty definitively stated, and C nails down the floating-point model more tightly than C++ does.
> 
> > arithmetic range and precision shall be at least as great as if the expression were being evaluated in the execution environment.
> 
> I don't think this is making a statement about value or correctness, is it? We just have to consider your same number of bits. At least that's how I read it.

I believe "The semantic rules for the evaluation of a constant expression are the same as for nonconstant expressions." is intended to tell you that you get the same results from the execution environment as you get from the translation environment. Whether that applies regardless of what those environments are (e.g., if they're different, as in cross-compiling), the standard is silent.

**Comment by jcranmer-intel** [comment]
C provides no firm guidance for the accuracy of math functions, save for explicitly disclaiming in Annex F that the recommended operations need not be correctly rounded. Annex F does lay on "as if" pretty heavily for runtime versus compile-time execution, but this is primarily around floating point exceptions. The inexact flag is explicitly unspecified, and whether or not the library functions respect rounding mode is implementation-defined.

My overall sense is that alignment between compile-time and runtime implementation is preferred, but there is some level of recognition that it's not fully practicable, with the intent that in the future, correctly-rounded implementations will exist that require everyone to agree on the precise result. And until that future world comes into being, it seems, the status quo is that people who care will understand that reliability will require careful application of compiler flags.

**Comment by andykaylor** [comment]
I just created a topic in Discourse to discuss this at the LLVM IR level.

The front end behavior is complicated by the fact that there are language rules that allow for different expression evaluation during static object initialization (see, for example, C11 F.8.4). However, I think apart from the language standard, there is a general user expectation of reproducible floating-point results. As such, I think unless we can point to a specific rule in some standard that explicitly allows numeric deviation, or the user has granted such permission (such as with -ffast-math), the default behavior of the compiler should be to produce consistent numeric results regardless of optimization level. Consistency across architectures is a different matter, and I don't think it can be assumed, but consistency for a single target (even in the case of cross-compilation) should be assumed.

This may prevent us from doing some things we'd like to do, such as constant folding function calls, but I think we (and our users) need to accept that. For instance, this rule prevents us from vectorizing function calls in some cases even when we have a vector implementation available that performs the requested operation, because we can't have vectorized code and scalar code producing different results.

By the same reasoning, I don't think using correctly rounded library implementations gets us out of this unless the compilation target is also promising to use correctly rounded implementations. Users, for the most part, care more about reproducibility than they do about absolute theoretical accuracy.
